cf-agent/package_module.c

```diff
@@ -0,0 +1,1604 @@
+                "Can not set up wrapper for module: %s", pm_name);
+            continue;
+        }
+
+        UpdateSinglePackageModuleCache(ctx, module_wrapper,
+                                       UPDATE_TYPE_INSTALLED, force_update);
+        UpdateSinglePackageModuleCache(ctx, module_wrapper,
+                                       force_update ? UPDATE_TYPE_LOCAL_UPDATES : 
+                                           UPDATE_TYPE_UPDATES,
+                                       force_update);
+
+        DeletePackageModuleWrapper(module_wrapper);
+    }
+    YieldGlobalPackagePromiseLock(package_lock);
+}
+
+PackagePromoseGlobalLock AcquireGlobalPackagePromiseLock(EvalContext *ctx)
+{
+    Bundle bundle = {.name = "package_global"};
+    PromiseType promise_type = {.name = "package_global",
+                               .parent_bundle = &bundle};
+    Promise pp = {.promiser = "package_global",
+                  .parent_promise_type = &promise_type};
+    
+    CfLock package_promise_global_lock;
+    
+    package_promise_global_lock =
+            AcquireLock(ctx, GLOBAL_PACKAGE_PROMISE_LOCK_NAME, VUQNAME, CFSTARTTIME,
+                        (TransactionContext) {.ifelapsed = VIFELAPSED,
+                                              .expireafter = VEXPIREAFTER},
+                        &pp, false);
+                        
+    return (PackagePromoseGlobalLock) {.g_lock = package_promise_global_lock,
+                                       .lock_ctx = ctx};
+}
+
+void YieldGlobalPackagePromiseLock(PackagePromoseGlobalLock lock)
+{
+    Bundle bundle = {.name = "package_global"};
+    PromiseType promise_type = {.name = "package_global",
+                               .parent_bundle = &bundle};
+    Promise pp = {.promiser = "package_global",
+                  .parent_promise_type = &promise_type};
+    
+    YieldCurrentLockAndRemoveFromCache(lock.lock_ctx, lock.g_lock,
+                                       GLOBAL_PACKAGE_PROMISE_LOCK_NAME, &pp);
+}
+
+static void ParseAndLogErrorMessage(const Rlist *data)
+{
+    for (const Rlist *rp = data; rp != NULL; rp = rp->next)
+    {
+        char *line = RlistScalarValue(rp);
+                   
+        if (StringStartsWith(line, "Error="))
+        {
+            Log(LOG_LEVEL_VERBOSE, "Caught error: %s", line + strlen("Error="));
+        }
+        else if (StringStartsWith(line, "ErrorMessage="))
+        {
+            Log(LOG_LEVEL_VERBOSE, "Caught error message: %s", 
+                line + strlen("ErrorMessage="));
+        }
+        else
+        {
+            Log(LOG_LEVEL_VERBOSE, "Caught unsupported error info: %s", line);
+        }
+    }
+}
+
+static int IsReadWriteReady(const IOData *io, int timeout_sec)
+{
+    fd_set  rset;
+    FD_ZERO(&rset);
+    FD_SET(io->read_fd, &rset);
+
+    struct timeval tv = {
+        .tv_sec = timeout_sec,
+        .tv_usec = 0,
+    };
+
+    int ret = select(io->read_fd + 1, &rset, NULL, NULL, &tv);
```

Comment:
I believe ``select`` does not work on file descriptors on Windows, only on sockets. We may have to consider a Windows specific section here.